cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/reader/TemplateSetConfigurationReader.java

```diff
@@ -0,0 +1,348 @@
+
+    if (!Files.exists(templateSetsDownloaded) && !Files.exists(templateSetsAdapted)) {
+      throw new InvalidConfigurationException(configRoot,
+          "Could not find a folder in which to search for the template set configuration file.");
+    } else {
+      if (Files.exists(templateSetsAdapted)) {
+        this.templateSetFiles.addAll(loadContextFilesAdapted(templateSetsAdapted));
+      }
+
+      if (Files.exists(templateSetsDownloaded)) {
+        this.templateSetFiles.addAll(loadContextFilesDownloaded(templateSetsDownloaded));
+      }
+
+      if (this.templateSetFiles.isEmpty()) {
+        throw new InvalidConfigurationException(configRoot, "Could not find any template set configuration file.");
+      }
+    }
+
+    this.templateSetRoot = configRoot;
+
+    readConfiguration();
+  }
+
+  /**
+   * Reads the template-set configuration.
+   */
+  protected void readConfiguration() {
+
+    // workaround to make JAXB work in OSGi context by
+    // https://github.com/ControlSystemStudio/cs-studio/issues/2530#issuecomment-450991188
+    final ClassLoader orig = Thread.currentThread().getContextClassLoader();
+    if (JvmUtil.isRunningJava9OrLater()) {
+      Thread.currentThread().setContextClassLoader(JAXBContext.class.getClassLoader());
+    }
+
+    this.templateSetConfigurations = new HashMap<>();
+
+    for (Path templateSetFile : this.templateSetFiles) {
+      try (InputStream in = Files.newInputStream(templateSetFile)) {
+        Unmarshaller unmarschaller = JAXBContext.newInstance(TemplateSetConfiguration.class).createUnmarshaller();
+
+        // Unmarshal without schema checks for getting the version attribute of the root node.
+        // This is necessary to provide an automatic upgrade client later on
+        Object rootNode = unmarschaller.unmarshal(in);
+        BigDecimal configVersion;
+        if (rootNode instanceof TemplateSetConfiguration) {
+          configVersion = ((TemplateSetConfiguration) rootNode).getVersion();
+          if (configVersion == null) {
+            throw new InvalidConfigurationException(templateSetFile,
+                "The required 'version' attribute of node \"templateSetConfiguration\" has not been set");
+          } else {
+            VersionValidator validator = new VersionValidator(Type.TEMPLATE_SET_CONFIGURATION, MavenMetadata.VERSION);
+            try {
+              validator.validate(configVersion.floatValue());
+            } catch (NotYetSupportedException e) {
+              // TODO
+            }
+          }
+        } else {
+          throw new InvalidConfigurationException(templateSetFile,
+              "Unknown Root Node. Use \"templateSetConfiguration\" as root Node");
+        }
+
+        // If we reach this point, the configuration version and root node has been validated.
+        // Unmarshal with schema checks for checking the correctness and give the user more hints to
+        // correct his failures
```

Comment:
I don't know what the general consensus is on in-code comments at the moment but i like this a lot, makes it a lot easier for others to understand what's going on